src/TextField/TextField.tsx

```diff
@@ -94,149 +105,168 @@ export const TextField: React.FC<Props> = ({
+            border: "solid 1px",
+            borderColor: error ? colors.red.base : colors.silver.darker,
+            "::placeholder": {
+              color: disabled ? colors.grey.lighter : colors.grey.light,
+              opacity: 1,
+            },
+            borderRadius: 4,
+            flex: 1,
+            height: size === "standard" ? 36 : size === "small" ? 28 : 42,
+            ...(size === "small"
+              ? typography.base.small
+              : typography.base.base),
+            marginRight: icon ? -30 : 0,
+            paddingLeft: icon ? 34 : size === "small" ? 8 : 12,
+            paddingRight: size === "small" ? 8 : 12,
+            width: "100%",
+            ":hover,  &[data-force-hover-state]": {
+              borderColor:
+                !disabled && !error
+                  ? colors.grey.light
+                  : error
+                  ? colors.red.base
+                  : colors.silver.darker,
+            },
+            ":focus, &[data-force-focus-state]": {
+              borderColor:
+                !disabled && !error
+                  ? colors.blue.light
+                  : error
+                  ? colors.red.base
+                  : colors.silver.darker,
+              outline: "none",
+            },
+          })
+        ),
+      };
+
+      return (
+        <div className={className}>
+          <label
+            className={cx(
+              css({
+                paddingBottom: 8,
+                ...typography.base.base,
+                fontWeight: 600,
+              })
+            )}
           >
-            {description}
-          </div>
-        )}
-        <div
-          css={{
-            marginTop: 8,
-            position: "relative",
-          }}
-        >
-          {icon && (
+            {label != null && <div css={{ marginBottom: 4 }}>{label}</div>}
+            {description != null && (
+              <div
+                css={{
+                  ...typography.base.base,
+                  color: colors.black.base,
+                }}
+              >
+                {description}
+              </div>
+            )}
             <div
               css={{
-                position: "absolute",
-                display: "inline-flex",
-                left: 12,
-                top: "50%",
-                transform: "translateY(-50%)",
+                marginTop: 8,
+                position: "relative",
               }}
             >
-              {icon}
+              {icon && (
+                <div
+                  css={{
+                    position: "absolute",
+                    display: "inline-flex",
+                    left: 12,
+                    top: "50%",
+                    transform: "translateY(-50%)",
+                  }}
+                >
+                  {icon}
+                </div>
+              )}
+
+              {React.isValidElement(inputAs)
+                ? React.cloneElement(inputAs, {
```

Comment:
I read the api documentation for [cloneElement](https://reactjs.org/docs/react-api.html#cloneelement) and [createElement](https://reactjs.org/docs/react-api.html#createelement) to understand what is happening here. From what I understand, `cloneElement` does a shallow copy of the props by default, so we need to merge the classNames explicitly. Will this have unintended consequences or semantics around class precedence that would require documentation?

The other case uses `createElement` without merging class names, since there are no props for `inputAs` when it is a string, such as `div` or `input`. It took me about 10 minutes to read through the React docs and understand it. I suggest you add a comment describing the cases here and why the prop merging is different, so hopefully that we don't need to context switch out of the source to understand the behavior

Reply:
>Will this have unintended consequences or semantics around class precedence that would require documentation

Great question. No, the order of the class names in `class="a b c"` has no effect on precedence. Precedence is determined solely by specificity, which has clearly defined rules in the css spec.

Reply:
>I suggest you add a comment describing the cases here and why the prop merging is different, so hopefully that we don't need to context switch out of the source to understand the behavior

I very much value the suggestion and am happy that you read through the docs before bringing up the talking point. 

First, understanding the difference between an element and component is critical. A component is the definition of something you will render, either a `React.FC` (functional component) or a class that extends `React.Component`. An element is the result of a render by either of two methods (note these two are identical; the first will be transpiled into the second by our build tools):

* jsx

    ```
    <a href="apollographql.com">link</a>
    ```

* `React.createElement`

    ```jsx
    React.createElement('a', {href: 'apollographql.com'}, 'link')
    ```

Note: `jsx` is transpiled into `React.createElement` for use in browsers.

---

As you know, `createElement` and `cloneElement` are very different things. `cloneElement` takes an element, meaning something you've already rendered, and augments it with additional props; hence the `classNames` merging. `createElement` takes either an intrinsic jsx element (like "div") or a component and renders it into a react element.

While this not trivial, I think it's more valuable to look at the React docs if someone wants to know how it works rather than explaining it in a comment. In my opinion, I value that you spent the time to read about `createElement` and `cloneElement` and I think you got far more out of that than any comment I could leave here. I also think it's better to read the docs because then in the next component where we do the same thing (Buttons, for example), you already know how they all work.